google/auth/external_account.py

```diff
@@ -210,6 +214,17 @@ def service_account_email(self):
                 return url[start_index:end_index]
         return None
 
+    @property
+    def external_account_id(self):
+        """Return the external account id.
+        When service account impersonation is used, it will return the service account email.
+        When no service account impersonation is used, it will return the principal.
```

Comment:
This is a little misleading as it can only be used by gcloud -- I don't think we should expose it.

Reply:
It's a public method, if it can be used by `gcloud`, it can be called by any other user client code. Since our library is just a dependency of `gcloud`

Reply:
Yes, exactly. It shouldn't be public. There is currently no way for someone to use these credentials and call token info. Only gcloud can do that.

Reply:
I'm little bit confused. If there is no public method, then how the library expose this to `gcloud`?

Reply:
I know it is possible technically, but seems not a good choice.
"Note that accessing a private method from outside the class is generally not recommended, as it violates the principle of encapsulation and can lead to unexpected behavior. It's better to rely on the public interface of the class to interact with its functionality."

Besides, don't we want to expose a public method for SDK users to have an insight of the token?
What I read from our token introspection proposal is:

```
Today, [GCP Security Token Service ](https://cloud.google.com/iam/docs/reference/sts/rest)(sts.googleapis.com) issues GCP access tokens. The contents of these tokens are opaque to clients. This means that the client can’t know anything about the content or structure of the token itself.
```

Reply:
Token introspection doesn't work without basic auth. SDK users can't call token info when using external account credentials. We can't expose a method that says this returns the principal because that won't work yet.

You can expose this on external account authorized user credentials, which are generated through gcloud auth application-default login and will be able to call token info.

Reply:
By basic auth I think you mean `client_id` and `client_secret`. How about the `external_account` throw an exception if there is no basic auth? And with a valid auth. We are offering this kind of operation only when there is an access token already.

Reply:
Expose `token_info_introspection` to be a public method when used with basic auth can return metadata of a token.
Throw an exception if there is:
1. No `token_info_url` configured
2. No basic auth with the external_account
3. No active token for the credential object

Expose `external_account_id` to be a property method. It will call visit cached value of `token_info_introspection` `username` filed if no `service_account_email` and throwing no error in any cases.